stk/utilities/utilities.py

```diff
@@ -1143,10 +1146,438 @@ def vector_theta(vector1, vector2):
+
+        # Regular expression for numbers.
+        nums = re.compile(r"[+-]?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?")
+        for line in reversed(self.output_string):
+            if '          | HOMO-LUMO GAP   ' in line:
+                value = nums.search(line.rstrip()).group(0)
+                break
+
+        return float(value)
+
+    def fermi_level(self):
+        """
+        Extracts Fermi-level energy in eV from xTB output.
+
+        Formatting based on latest version of xTB (190418)
+        Example line:
+        "             Fermi-level           -0.3159871 Eh
+                   -8.5984 eV"
+
+        Returns
+        -------
+        :class:`float`
+            Fermi-level in eV.
+        """
+        value = None
+
+        # Regular expression for numbers.
+        nums = re.compile(r"[+-]?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?")
+        for line in reversed(self.output_string):
+            if '             Fermi-level        ' in line:
+                part2 = line.split('Eh')
+                value = nums.search(part2[1].rstrip()).group(0)
+                break
+
+        return float(value)
+
+    def qonly_dipole_moment(self):
+        """
+        Extracts `q only` dipole moment in Debye from xTB output.
+
+        Formatting based on latest version of xTB (190418)
+        Example line:
+        " q only:       -0.033      -0.081      -0.815"
+
+        Returns
+        -------
+        :class:`list` of :class:`float`
+            Components of dipole moment as ``[`x`, `y`, `z`]`` in a
+            list of length 3.
+
+        """
+        value = None
+
+        sample_set = []
+        for i, line in enumerate(self.output_string):
+            if 'molecular dipole:' in line:
+                sample_set = self.output_string[i+2].rstrip()
+
+        # get values from line
+        if 'q only:' in sample_set:
+            x, y, z = [i for i in sample_set.split(':')[1].split(' ')
+                       if i != '']
+
+        value = [float(x), float(y), float(z)]
+
+        return value
+
+    def full_dipole_moment(self):
+        """
+        Extracts `full` dipole moment in Debye from xTB output.
+
+        Formatting based on latest version of xTB (190418)
+        Example line:
+        "   full:       -0.684       0.122      -1.071       3.245"
+
+        Returns
+        -------
+        :class:`list` of :class:`float`
+            Components of dipole moment and total magnitude as
+            ``[`x`, `y`, `z`, `total`]`` in a list of length 4.
+
+        """
+        value = None
+
+        sample_set = []
+        for i, line in enumerate(self.output_string):
+            if 'molecular dipole:' in line:
+                sample_set = self.output_string[i+3].rstrip()
+
+        # get values from line
+        if 'full:' in sample_set:
+            samp = sample_set.split(':')[1].split(' ')
+            x, y, z, m = [i for i in samp if i != '']
+
+        value = [float(x), float(y), float(z), float(m)]
```

Comment:
again, use the list comprehension more efficiently here and get rid of comparison to emptyu string.